### DISCUSSION TITLE
Support for other base containers

### DISCUSSION THREAD

**Discussion by psychemedia** [mmh352/ou-container-builder#23] [Ideas]
*Starting this as a discussion because there could be a lot of to-ing an fro-ing and you probably won't like it because it veers towards the very general;-)*

The current set up uses a hardwired base container (`python:3.9-slim-buster`).

Should this be definable via a `ContainerConfig.yaml` element (`image:`?) with default `python:3.9-slim-buster`?

In the general case, the use of single, hard wired base container may be limiting in at least two ways:

1. To create a container that runs a specific application, it may be easier to take a container off-the-shelf and top it up with OU customisations. For example, a module that wants to run R notebooks or RStudio. It would be easier to just pull an R environment image off the shelf.

2. Several modules may benefit from using an image with 70% the same basis image, or a module may be happy with a solid base container and then a few custom layers each presentation. In this case, it might make sense for them to be able just declare which base container they want to use (so a `ContainerConfig.base.yaml` builds their base container, then another `ContainerConfig.topup.yaml` pulls that image and tops up).

However, the deployment model does require each container to support a python environment to run the `jupyter-server` and optionally a `jupyter-server-proxy` as a harness.

## Build Routes

Does it make sense to consider two build routes (which may simplify to one?):

1. build on top of a Python container as currently happens;
2. use a multistage build to harness and then copy the harness requirements into a custom specified base container. For example:

```
FROM python:3.9-slim-buster

# Install and run script to set up Jupyter server etc
# Alternatively, this could be pre-packaged in a harness container
# (for example, `FROM ou-container-builder-harness`)
# with the environment in a known location

FROM $CUSTOM_IMAGE 

# Copy over the harness environment that runs the `jupyter-server` etc

# Add in the rest of the ou-container-build steps specified in `ContainerCongfig.yaml` here
```

Using an ou-container-build built "stack" cotnainer as the initial step in a multi-build would be complicated by trying to figure out what to copy over, unless everything we installed into a known location/locations.

**Comment by mmh352** [comment]
There will always be edge-cases where full control over the images is needed, but that is outside of the scope for this project. The ``ou-container-builder`` is designed as a tool that makes it very simple to build the majority of scenarios. The developer is not meant to have to think about base containers, the structure of the Dockerfile, or how the directories are set up. Their focus should be on what they want the container to have. You want a Jupyter notebook container with some files - set the type to "jupyter-notebook", specify the files to copy, build and done. You want a Jupyter notebook with an R kernel - set the type to "jupyter-notebook", add the "jupyter-r" pack, specify the files to copy, build and done. You need some unusual setup based on a container somebody else has built that you can't look into - you are on your own.

The reason for the hard-wired base container is that it guarantees the specific Python version, which allows for testing of the core Python components (jupyterhub, jupyter_server) against that Python version. This simplifies problem solving, because if there are issues, then they are caused by the functionality being layered on top and this can easily be stripped out until the bug is found and then fixed. Supporting arbitrary base images in the container means that the ``ou-container-builder`` could no longer guarantee that the basic Jupyter level of functionality will work or, in a worst-case scenario, that the basic Jupyter packages will even install correctly.

I do like multi-stage builds, but basically the architectural design of the ``ou-container-builder`` ignores the advantages they bring in reusability in favour of making the basic scenario easy to use for the developer. Instead of re-using images, additional "packs" should be added to the ``ou-container-builder`` to simplify re-use.

**Reply by psychemedia** [reply]
Hmm.. so, why couldn't you just create a Jupyter harness with a known Py version and self-contain the jupyter server?

(Actually, I can maybe see one issue, which is what py environment do you install packages in? The server environment (which would be required for server proxy applications, I think), or the default Jupyter Python kernel (which may be separate).(

Hmmm... I'm wondering now... do you always get the Py environment that runs the `jupyter-server` as a kernel? Or can you suppress it use or define another Python environment kernel as the default?

**Reply by mmh352** [reply]
Of course this is all possible, but it is a lot more work, requires more testing, and achieves exactly the same as the current system already does. Until I see an example of a simple to moderate complexity container that cannot be built within the current setup, I just don't see any reason for a complex solution to a hypothetical problem.

**Reply by psychemedia** [reply]
Ok...  You're the one who signed up to be the image designer ;-) (I just think back to all the joy fun times I've had in the past negotiating dependency hell trying to figure out why a particular geo container, or music container, or chemistry container wouldn't build properly if an off-the-shelf version wasn't available and "just worked" ;-) Or why I wasted a chunk of time yesterday not being able to build the OpenRefine container when every other Dockerfile I've ever created to install it has worked trivially;-)

LaTeX configs can also be a faff at times...

**Reply by mmh352** [reply]
Oh, I'm sure that that will suck up plenty of work. But debugging a container build locally is still much easier than debugging the deployed version of something where somebody has done something and is running it in some way. That way lies madness.
